### PR TITLE
ENT-284: Refine SAP SuccessFactors course export flow

### PR DESCRIPTION
**Pull request by haikuginger**
**Description:**

This pull request finalizes a couple features to do with the SAP course export feature. Specifically:

1. Retrieve the course language and use it in the submission
2. Create a course track selection URL for the main link, rather than the marketing URL
3. Send up to 1000 courses at once to SAP

**JIRA:** [ENT-284](https://openedx.atlassian.net/browse/ENT-284)

**Dependencies:** None

**Merge deadline:** 10 April 2017

**Installation instructions:**
Requires:
- Course Discovery
    - At least one course
    - At least one catalog containing at least one course
- Enterprise customer
    - Attached to catalog required above
- Configured SapSuccessFactorsGlobalConfiguration (see document provided offline)
- Configured SapSuccessFactorsEnterpriseCustomerConfiguration (see document provided offline)
    - Link to our existing EnterpriseCustomer

**Testing instructions:**
1. Run the following (where `staff` is a local superuser with permissions to the course catalog):
    ```sh
    ./manage.py lms transmit_courseware_data --catalog_user staff --settings=devstack
    ```
2. In the stdout logging of your devstack, observe the following facts:
    - The `locale` key under `description` and `title` should (roughly) match the language entered for the course in the Course Discovery stack, unless the language selected there isn't available. In that case, it should be `English`. Change this selection around a bit, and rerun the command to be sure it tracks expected behavior.
    - Under `content`, the single present item should have a `launch_url` that corresponds to the course track selection page, and which matches the domain of the site linked to your EnterpriseCustomer. Scheme will be `https` by default, unless you set the `SUCCESSFACTORS_COURSE_EXPORT_DEFAULT_URL_SCHEME` environment variable before running the command. Example:
         ```
        https://example.com/course_modes/choose/course-v1:edX+fancy+demo123/
        ```

**Merge checklist:**

- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:**

1. Because I was sick of seeing three uncovered lines, I also wrote tests for them. 100% coverage!

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=79&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

File: integrated_channels/sap_success_factors/transmitters/courses.py [hunk 1]
```diff
@@ -18,16 +18,18 @@ class SuccessFactorsCourseTransmitter(SuccessFactorsTransmitterBase):
...
         Args:
-            payload (SapCourseExporter): The OCN course exporter object to send to SAP SuccessFactors
+            serialized_payload (bytes): A set of bytes containing a page's worth of data
+
+        Returns:
+            code: An integer status for the HTTP request
+            body: The server's response body
         """
-        serialized_payload = payload.get_serialized_data()
         LOGGER.info(serialized_payload)
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Will this `LOGGER` statement remain after the PR has merged?

File: integrated_channels/sap_success_factors/transmitters/courses.py [hunk 2]
```diff
@@ -18,16 +18,18 @@ class SuccessFactorsCourseTransmitter(SuccessFactorsTransmitterBase):
...
-
         try:
             code, body = self.client.send_course_import(serialized_payload)
             LOGGER.debug('Successfully sent course metadata for Enterprise Customer {}'.
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Will this `LOGGER` statement remain after the PR has merged?

File: integrated_channels/sap_success_factors/transmitters/courses.py [hunk 3]
```diff
@@ -38,7 +40,29 @@ def transmit(self, payload):
...
+    def transmit(self, payload):
+        """
+        Send a course data import call to SAP SuccessFactors using the client.
+
+        Args:
+            payload (SapCourseExporter): The OCN course exporter object to send to SAP SuccessFactors
+        """
+        total_transmitted = 0
+        errors = []
+        codes = []
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Nit: `status_codes` would be a better name to use.

File: integrated_channels/sap_success_factors/transmitters/courses.py [hunk 4]
```diff
@@ -38,7 +40,29 @@ def transmit(self, payload):
             LOGGER.error('Failed to send course metadata for Enterprise Customer {}\nError Message {}'.
                          format(self.enterprise_configuration.enterprise_customer.name, body))
 
-        error_message = body if code >= 400 else ''
+        return code, body
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Nit: `code` -> `status_code`

File: integrated_channels/sap_success_factors/transmitters/courses.py [hunk 5]
```diff
@@ -18,16 +18,18 @@ class SuccessFactorsCourseTransmitter(SuccessFactorsTransmitterBase):
...
         """
-        Send a course data import call to SAP SuccessFactors using the client.
+        SAPSuccessFactors can only send 1000 items at a time, so this method sends one "page" at a time.
 
         Args:
-            payload (SapCourseExporter): The OCN course exporter object to send to SAP SuccessFactors
+            serialized_payload (bytes): A set of bytes containing a page's worth of data
+
+        Returns:
+            code: An integer status for the HTTP request
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: code -> status_code

File: integrated_channels/sap_success_factors/transmitters/courses.py [hunk 6]
```diff
@@ -18,16 +18,18 @@ class SuccessFactorsCourseTransmitter(SuccessFactorsTransmitterBase):
...
+            body: The server's response body
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: body -> response_body

Comment (mattdrayer) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Also please clarify who the "server" is in this case -- eg, the "SAP SuccessFactors server"

File: integrated_channels/sap_success_factors/utils.py [hunk 7]
```diff
@@ -5,32 +5,164 @@
...
+        """
+        Return serialized blocks of data representing the courses to be POSTed, 1000 at a time.
+
+        Yields:
+            bytes: JSON-serialized course metadata structure
+            int: Number of records in this batch
+        """
+        page_length = 1000
+        location = 0
+        while (location < len(self.courses)) or not location:
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/utils.py]: Performance nit -- you can set len(self.courses) as a variable prior to the loop and then you will not have to re-evaluate the length for every iteration -- this will be helpful especially if we are talking about 1000s of courses.

File: integrated_channels/sap_success_factors/utils.py [hunk 8]
```diff
@@ -77,6 +209,20 @@ def get_serialized_data(self):
     }
 
 
+def get_course_track_selection_url(enterprise_customer, course_id):
+    """
+    Given a course ID, and EnterpriseCustomer, craft a URL that links to the track selection page for that course.
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/utils.py]: Nit: Technically it's "Given an EnterpriseCustomer and a course ID..."

File: integrated_channels/sap_success_factors/utils.py [hunk 9]
```diff
@@ -93,3 +239,27 @@ def current_time_is_in_interval(start, end):
     interval_start = parse_lms_api_datetime(start or UNIX_MIN_DATE_STRING)
     interval_end = parse_lms_api_datetime(end or UNIX_MAX_DATE_STRING)
     return interval_start <= timezone.now() <= interval_end
+
+
+def parse_language_code(code):
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/utils.py]: "Parse" doesn't really explain what is going on inside this function -- if we're doing a transform, then let's be clear about it -- something like `transform_language_code` or `transform_language_code_to_name` would work.

File: integrated_channels/sap_success_factors/utils.py [hunk 10]
```diff
@@ -93,3 +239,27 @@ def current_time_is_in_interval(start, end):
...
+    """
+    Transform an ISO language code (for example, en-us) into a language name
+    in the format read by SuccessFactors.
+    """
+    if code is None:
+        return 'English'
+    components = code.split('-')
+    language_code = components[0]
+    if 3 > len(components) > 1:
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/utils.py]: I'm trying to understand why we would not just say "2" here -- must be missing something 😄

File: integrated_channels/sap_success_factors/utils.py [hunk 11]
```diff
@@ -93,3 +239,27 @@ def current_time_is_in_interval(start, end):
...
     interval_end = parse_lms_api_datetime(end or UNIX_MAX_DATE_STRING)
     return interval_start <= timezone.now() <= interval_end
+
+
+def parse_language_code(code):
+    """
+    Transform an ISO language code (for example, en-us) into a language name
+    in the format read by SuccessFactors.
+    """
+    if code is None:
```
Comment (mattdrayer) [inline integrated_channels/sap_success_factors/utils.py]: code -> language_code

Comment (mattdrayer) [inline integrated_channels/sap_success_factors/utils.py]: Ah, forget it -- you are using that below -- this can stay as-is

File: tests/test_management.py [hunk 12]
```diff
@@ -92,12 +93,15 @@ def test_transmit_courseware_task_success(fake_client, fake_catalog_client, capl
...
 
+    track_selection_reverse_mock.return_value = '/course_modes/choose/course-v1:edX+DemoX+Demo_Course/'
+
     caplog.set_level(logging.INFO)
 
     UserFactory(username='C-3PO')
     enterprise_customer = EnterpriseCustomerFactory(
         catalog=1,
         name='Veridian Dynamics',
+        site__domain='supersuccessful.com'
```
Comment (mattdrayer) [inline tests/test_management.py]: Prolly should stick with `example.com` -- this actually appears to be a registered domain name although nothing renders.

File: tests/test_models.py [hunk 13]
```diff
@@ -658,7 +658,7 @@ def test_string_conversion(self, method):
         )
         self.assertEqual(method(customer_branding_config), expected_str)
 
-    def test_logo_path(self):
+    def test_logo_path_does_not_exist(self):
```
Comment (mattdrayer) [inline tests/test_models.py]: Drive-by refactor/fixup?

**Comment by mattdrayer** [comment]
Initial pass complete!

File: tests/test_models.py [hunk 14]
```diff
@@ -670,9 +670,29 @@ def test_logo_path(self):
...
+            enterprise_customer=EnterpriseCustomerFactory(),
+            logo=file_mock
+        )
+
+        storage_mock = mock.MagicMock(spec=Storage, name="StorageMock")
+        storage_mock.exists.return_value = True
         with mock.patch("django.core.files.storage.default_storage._wrapped", storage_mock):
             path = logo_path(branding_config, branding_config.logo.name)
             self.assertEqual(path, "enterprise/branding/1/1_logo.png")
+            storage_mock.delete.assert_called_once_with("enterprise/branding/1/1_logo.png")
```
Comment (itsjeyd) [inline tests/test_models.py]: @haikuginger Looks like the only thing that varies between `test_logo_path_does_not_exist` and `test_logo_path_exists` is `storage_mock.exists.return_value`. You could DRY these tests out by turning them into data-driven tests.

File: integrated_channels/sap_success_factors/utils.py [hunk 15]
```diff
@@ -77,6 +209,20 @@ def get_serialized_data(self):
...
+    Args:
+        enterprise_customer (EnterpriseCustomer): The EnterpriseCustomer that a URL needs to be built for
+        course_id (str): The string identifier of the course in question
+    """
+    netloc = enterprise_customer.site.domain
+    scheme = COURSE_URL_SCHEME
+    path = reverse('course_modes_choose', args=[course_id])
+    return urlunparse((scheme, netloc, path, None, None, None))
+
+
```
Comment (itsjeyd) [inline integrated_channels/sap_success_factors/utils.py]: @haikuginger Should we introduce a dedicated test for this function?

File: integrated_channels/sap_success_factors/utils.py [hunk 16]
```diff
@@ -5,32 +5,164 @@
...
-class SapCourseExporter(BaseCourseExporter):
+class SapCourseExporter(BaseCourseExporter):  # pylint: disable=abstract-method
     """
     Class to provide data transforms for SAP SuccessFactors course export task.
     """
 
     def __init__(self, user, plugin_configuration):
         super(SapCourseExporter, self).__init__(user, plugin_configuration)
 
-    def get_serialized_data(self):
```
Comment (itsjeyd) [inline integrated_channels/sap_success_factors/utils.py]: @haikuginger I think you might also need to rename this to `get_serialized_data_blocks` in the super class.

File: integrated_channels/sap_success_factors/transmitters/courses.py (hunk 1 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: It isn't new, so I don't see why not. Would you like it removed?

File: integrated_channels/sap_success_factors/transmitters/courses.py (hunk 2 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: As above.

File: integrated_channels/sap_success_factors/transmitters/courses.py (hunk 3 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Noted.

File: integrated_channels/sap_success_factors/transmitters/courses.py (hunk 4 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Noted.

File: integrated_channels/sap_success_factors/transmitters/courses.py (hunk 5 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Noted.

File: integrated_channels/sap_success_factors/transmitters/courses.py (hunk 6 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: Both points noted.

File: integrated_channels/sap_success_factors/utils.py (hunk 7 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: Good call - `len(self.courses)` should be a nice constant-time expression, but there's no reason to have it inside the loop.

File: integrated_channels/sap_success_factors/utils.py (hunk 8 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: Fair. Noted.

File: integrated_channels/sap_success_factors/utils.py (hunk 9 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: Sounds good!

File: integrated_channels/sap_success_factors/utils.py (hunk 10 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: Ah. Yeah, that'd work. 😅

File: tests/test_management.py (hunk 12 above)
Comment (haikuginger) [inline tests/test_management.py]: Fair enough! Noted.

File: integrated_channels/sap_success_factors/utils.py [hunk 17]
```diff
@@ -93,3 +239,27 @@ def current_time_is_in_interval(start, end):
...
+        country_code = '_'
+    else:
+        raise ValueError('Language codes may only have up to two components. Could not parse: {}'.format(code))
+
+    if language_code not in SUCCESSFACTORS_OCN_LANGUAGE_CODES:
+        LOGGER.warning('Language "%s" is not supported by SuccessFactors. Sending English by default.', language_code)
+        return 'English'
+
+    language_family = SUCCESSFACTORS_OCN_LANGUAGE_CODES[language_code]
+    return language_family.get(country_code, language_family['_'])
```
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: I can definitely make the code take up more lines. 😉

No, we don't need to be concerned. We already handle the case above where the language code isn't in `SUCCESSFACTORS_OCN_LANGUAGE_CODES` by falling back to English. Here, either the country code will be found in the language family (meaning that there's a dialect for that country that has a specific defined name for SAPSF ("Canadian English" for `en-ca`)), or it won't be, and we'll fall through to the base language name (defined by `_`).

File: tests/test_models.py (hunk 13 above)
Comment (haikuginger) [inline tests/test_models.py]: 100% coverage, baby!

File: tests/test_models.py (hunk 14 above)
Comment (haikuginger) [inline tests/test_models.py]: Good call.

File: integrated_channels/sap_success_factors/utils.py (hunk 16 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: I don't think we want to here. `get_serialized_data` is still the canonical form for integrated channels in general; we just have to do `get_serialized_data_blocks` here because we need to page through a given number of elements at a time.

File: integrated_channels/sap_success_factors/utils.py (hunk 15 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: Couldn't hurt!

**Comment by mattdrayer** [comment]
Updates LGTM -- no further changes required from me -- @brittneyexline please take a pass thru when you have a chance, thanks -- 👍

File: integrated_channels/sap_success_factors/utils.py (hunk 16 above)
Comment (itsjeyd) [inline integrated_channels/sap_success_factors/utils.py]: @haikuginger OK. Does that mean that subclasses of `BaseCourseExporter` don't *have to* implement [`get_serialized_data`](https://github.com/edx/edx-enterprise/blob/master/integrated_channels/integrated_channel/course_metadata.py#L113-L117)?

File: integrated_channels/sap_success_factors/utils.py (hunk 15 above)
Comment (itsjeyd) [inline integrated_channels/sap_success_factors/utils.py]: @haikuginger But you're not going to do that now? 😄

File: integrated_channels/sap_success_factors/utils.py (hunk 16 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: They don't _have_ to. It's the canonical method, but obviously doesn't work here for obvious reasons. I'd say it's something that allows for components to be developed separately and synchronously, but still have a commitment to a common API. In this case, we've already done that part, so it's okay to deviate from it so that we can support the feature properly.

File: integrated_channels/sap_success_factors/utils.py (hunk 15 above)
Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: Whoops, thought you were pointing to the other function! Will add a couple tests for this as well. :)

**Comment by haikuginger** [comment]
@itsjeyd, tests updated to include the one you _actually_ requested. 😜

**Comment by itsjeyd** [comment]
@haikuginger That's great, thanks! 😄

File: integrated_channels/sap_success_factors/utils.py [hunk 18]
```diff
@@ -5,32 +5,165 @@
...
-        return json.dumps(final_structure, sort_keys=True).encode('utf-8')
+    def get_serialized_data_blocks(self):
+        """
+        Return serialized blocks of data representing the courses to be POSTed, 1000 at a time.
+
+        Yields:
+            bytes: JSON-serialized course metadata structure
+            int: Number of records in this batch
+        """
+        page_length = 1000
```
Comment (brittneyexline) [inline integrated_channels/sap_success_factors/utils.py]: Should this be a constant in the class?

Comment (haikuginger) [inline integrated_channels/sap_success_factors/utils.py]: Good call!

**Comment by haikuginger** [comment]
Squashed, checklist, rebased, etc, from 5ad8b1f.
